app/controllers/SupplementaryUnitController.scala

```diff
@@ -61,34 +61,37 @@ class SupplementaryUnitController @Inject() (
         case None        => form
         case Some(value) => form.fill(value)
       }
-
-      val result = for {
-        query              <- request.userAnswers.get(RecordCategorisationsQuery)
-        categorisationInfo <- query.records.get(recordId)
-      } yield {
-        val measurementUnit = categorisationInfo.measurementUnit.getOrElse("")
-        val submitAction    = routes.SupplementaryUnitController.onSubmit(mode, recordId)
-        Ok(view(preparedForm, mode, recordId, measurementUnit, submitAction))
+      val catInfo      = request.userAnswers.get(LongerCategorisationDetailsQuery(recordId)) match {
+        case Some(catInfo) => Some(catInfo)
+        case _             => request.userAnswers.get(CategorisationDetailsQuery(recordId))
       }
-
-      result.getOrElse(Redirect(routes.JourneyRecoveryController.onPageLoad().url))
+      catInfo
+        .map { categorisationInfo =>
+          val measurementUnit = categorisationInfo.measurementUnit.getOrElse("")
+          val submitAction    = routes.SupplementaryUnitController.onSubmit(mode, recordId)
+          Ok(view(preparedForm, mode, recordId, measurementUnit, submitAction))
+        }
+        .getOrElse(Redirect(routes.JourneyRecoveryController.onPageLoad().url))
     }
 
   def onSubmit(mode: Mode, recordId: String): Action[AnyContent] =
     (identify andThen profileAuth andThen getData andThen requireData).async { implicit request =>
-      val onSubmitAction: Call = routes.SupplementaryUnitController.onSubmit(mode, recordId)
       form
         .bindFromRequest()
         .fold(
           formWithErrors => {
-            val result = for {
-              query              <- request.userAnswers.get(RecordCategorisationsQuery)
-              categorisationInfo <- query.records.get(recordId)
-            } yield {
-              val measurementUnit = categorisationInfo.measurementUnit.getOrElse("")
-              Future.successful(BadRequest(view(formWithErrors, mode, recordId, measurementUnit, onSubmitAction)))
+            val catInfo = request.userAnswers.get(LongerCategorisationDetailsQuery(recordId)) match {
+              case Some(catInfo) => Some(catInfo)
+              case _             => request.userAnswers.get(CategorisationDetailsQuery(recordId))
             }
-            result.getOrElse(Future.successful(Redirect(routes.JourneyRecoveryController.onPageLoad().url)))
+
+            catInfo
+              .map { categorisationInfo =>
+                val measurementUnit = categorisationInfo.measurementUnit.getOrElse("")
```

Comment:
when this is "allowed" to be empty the option should be None, why are we doing getOrElse("") here? :)

Reply:
It seems to be driven by the view. Tempted to change it. I think if the view should decide what it displays rather than the controller having to care :)

Reply:
Changed this now so the view deals with the option :) Also means its in one place rather than 4!